pennylane/ops/mid_circuit_measure.py

```diff
@@ -0,0 +1,291 @@
+
+        .. code-block:: python
+
+            df3jff4t=0,f93fjdj3=0 => 3.4,100
+            df3jff4t=0,f93fjdj3=1 => 3.4,67
+            df3jff4t=1,f93fjdj3=0 => 1,100
+            df3jff4t=1,f93fjdj3=1 => 1,67
+
+        (note the uuids in the example represent distinct measurements of different qubit.)
+
+        """
+        if isinstance(other, MeasurementDependantValue):
+            if self._dependent_on == other._dependent_on:
+                return MeasurementDependantValue(
+                    self._dependent_on,
+                    self._zero_case._merge(other._zero_case),
+                    self._one_case._merge(other._one_case),
+                )
+            if self._dependent_on < other._dependent_on:
+                return MeasurementDependantValue(
+                    self._dependent_on,
+                    self._zero_case._merge(other),
+                    self._one_case._merge(other),
+                )
+            return MeasurementDependantValue(
+                other._dependent_on,
+                self._merge(other._zero_case),
+                self._merge(other._one_case),
+            )
+        return MeasurementDependantValue(
+            self._dependent_on,
+            self._zero_case._merge(other),
+            self._one_case._merge(other),
+        )
+
+    def _transform_leaves_inplace(self, fun: Callable):
+        """
+        Transform the leaves of a MeasurementDependantValue with `fun`.
+        """
+        self._zero_case._transform_leaves_inplace(fun)
+        self._one_case._transform_leaves_inplace(fun)
+
+
+# pylint: disable=too-few-public-methods,protected-access
+class _Value(Generic[T]):
+    """
+    Leaf node for a MeasurementDependantValue tree structure.
+    """
+
+    __slots__ = ("_values",)
+
+    def __init__(self, *values):
+        self._values = values
+
+    def _merge(self, other):
+        """
+        Works with MeasurementDependantValue._merge
+        """
+        if isinstance(other, MeasurementDependantValue):
+            return MeasurementDependantValue(
+                other._dependent_on, self._merge(other._zero_case), self._merge(other._one_case)
+            )
+        if isinstance(other, _Value):
+            return _Value(*self._values, *other._values)
+        return _Value(*self._values, other)
+
+    def _transform_leaves_inplace(self, fun):
+        """
+        Works with MeasurementDependantValue._transform_leaves
+        """
+        self._values = (fun(*self._values),)
+
+    @property
+    def values(self):
+        """Values this Leaf node is holding."""
+        if len(self._values) == 1:
+            return self._values[0]
+        return self._values
+
+
+def if_then(expr: MeasurementDependantValue[bool], then_op: Type[Operation]):
+    """
+    Run an operation conditionally on the outcome of mid-circuit measurements.
+
+    .. code-block:: python
+
+        m0 = qml.mid_measure(0)
+        qml.if_then(m0, qml.RZ)(1.2, wires=1)
```

Comment:
Would `qml.if_then(m_0 + m_1, qml.RX)(math.pi, wires=2)` be valid? If so, how could we check for each option of `m_0 + m_1`? Could it make sense to make sense for `qml.if_then` to accept a boolean as the first argument?